shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -47,71 +68,116 @@ public String getCommandName() {
     return "persist";
   }
 
+  @Override
+  public Options getOptions() {
+    return new Options().addOption(PARALLELISM_OPTION);
+  }
+
   @Override
   public void validateArgs(CommandLine cl) throws InvalidArgumentException {
     CommandUtils.checkNumOfArgsNoLessThan(this, cl, 1);
   }
 
   @Override
-  protected void runPlainPath(AlluxioURI plainPath, CommandLine cl)
-      throws AlluxioException, IOException {
-    persist(plainPath);
+  public String getUsage() {
+    return "persist [-p|--parallelism <#>] <path> [<path> ...]";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Persists files or directories currently stored only in Alluxio to the UnderFileSystem.";
   }
 
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
+    int parallelism = 4;
+    if (cl.hasOption(PARALLELISM_OPTION.getLongOpt())) {
+      String parellismOption = cl.getOptionValue(PARALLELISM_OPTION.getLongOpt());
+      parallelism = Integer.parseInt(parellismOption);
+    }
     String[] args = cl.getArgs();
+    List<AlluxioURI> candidateUris = new ArrayList<>();
     for (String path : args) {
-      AlluxioURI inputPath = new AlluxioURI(path);
-      runWildCardCmd(inputPath, cl);
+      candidateUris.addAll(FileSystemShellUtils.getAlluxioURIs(mFileSystem, new AlluxioURI(path)));
+    }
+    final BlockingQueue<AlluxioURI> toPersist = new LinkedBlockingQueue<>();
+    for (AlluxioURI uri : candidateUris) {
+      queuePersist(mFileSystem.getStatus(uri), toPersist);
+    }
+    System.out.println("Found " + toPersist.size() + " files to persist.");
+    ExecutorService service =
+        Executors.newFixedThreadPool(parallelism, ThreadFactoryUtils.build("persist-cli-%d", true));
```

Comment:
Could we use `int parallelism = Math.min(toPersist.size(), 4)` here?

Reply:
Would `min(toPersist.size(), parallelism)` make more sense?

Reply:
Yes! That's better. we also need to deal with `toPersist.size() == 0` circumstances.